### ISSUE TITLE
disagreement-gated row-major preference in `TensorIterator::reorder_dimensions`

### ISSUE THREAD

**Issue by HU-qingqing** [pytorch/pytorch#191186]
### 🚀 The feature, motivation and pitch

# Proposal: disagreement-gated row-major preference in `TensorIterator::reorder_dimensions`

## Summary

`TensorIterator::reorder_dimensions` currently lets the **first decidable operand**
(stride non-zero on both dims of the pair) determine the swap for each dim pair,
via `should_swap`. When operands disagree on the swap direction, the tie is broken
by add order, which has no relation to memory denseness or layout friendliness.

This proposal: when multiple decidable operands *disagree* on a dim pair,
preferring the no-swap outcome (which keeps `perm_` closer to the identity
`[n-1, ..., 0]`, i.e. row-major). When operands agree, or when only one operand is
decidable, behavior is unchanged.

This is a per-dim-pair, conditionally-activated relaxation of the existing
`enforce_linear_iteration_` flag — it kicks in only on disagreement, so all
single-operand cases (unary, in-place) and all same-layout cases (including
channels-last propagation) are untouched.

## Motivation

Under the current `should_swap` rule, `c = a + b` and `c = b + a` can produce
outputs with **different layouts**: the first decidable operand per dim pair
returns `+1/-1` and the rest are never consulted, so per-dim-pair precedence is
purely a function of add order. When the two operands have incompatible layouts
(one row-major, one permuted), the output is row-major or permuted depending on
which operand is written first — even though elementwise addition is
commutative.

Concrete example:

```python
import torch
device = "cuda"
# Input 0: padded and permuted (channels-last-like ordering). Strides are
# not all ascending, so should_swap returns a non-zero verdict on some dim
# pairs.
tensor_0 = torch.empty_strided(
    [1, 6, 256, 64, 176],
    [22978560, 3829760, 1, 45056, 256],   # has padding gap, not dense
    device=device, dtype=torch.float32)
# Input 1: dense row-major contiguous.
tensor_1 = torch.empty_strided(
    [1, 6, 256, 64, 176],
    [17301504, 2883584, 11264, 176, 1],   # row-major, is_non_overlapping_and_dense() == True
    device=device, dtype=torch.float32)

c01 = tensor_0 + tensor_1     # should_swap consults tensor_0 first → inherits tensor_0's perm
c10 = tensor_1 + tensor_0     # should_swap consults tensor_1 first → inherits tensor_1's perm (row-major)
# c01.strides() != c10.strides(), even though + is commutative.
```

The downstream op re-runs `reorder_dimensions` independently, so whichever
layout was chosen at the `+` propagates forward; the layout offered by
*one* of the operands is lost when the other operand happens to come first.

The existing rule produces a layout that is "inherited by default" rather
than "chosen because it is friendlier". `should_swap` only compares relative
stride magnitudes between two dims of a single operand and has no notion of
overall memory denseness or of disagreement between operands.

## Proposed semantics (one-line summary)

> When decidable operands **agree** on a dim pair, the agreed verdict stands
> (existing behavior, including channels-last propagation).
> When decidable operands **disagree**, return `0` (no-swap) — biasing `perm_`
> back toward identity `[n-1, ..., 0]` (row-major).
> When only one operand is decidable, or none is, behavior is unchanged.

This makes `c = a + b` and `c = b + a` produce the same output layout whenever
the two operands disagree (both yield row-major), restoring commutativity at
the layout level for the cases above. Same-layout operands (including NHWC +
NHWC) still propagate their shared layout, because they agree on every dim
pair.

## Prior art: `enforce_linear_iteration_`

`TensorIteratorConfig::enforce_linear_iteration()` (declared in
`aten/src/ATen/TensorIterator.h:881`) forces `perm_ = [n-1, ..., 0]`
unconditionally (see `TensorIterator.cpp:247-250`):

```cpp
if (enforce_linear_iteration_) {
  permute_dimensions(perm_);
  return;
}
```

It is defaulted to `false`, but `nonzero_out_cpu` and `masked_scatter_`
opt in because their downstream stubs require C-order traversal. The
default-off stance reflects that *unconditional* row-major forcing is
too aggressive for general elementwise ops -- it breaks layout
propagation and harms unary / in-place permuted views.

This proposal is the *conditionally-activated* form: bias toward C-order
(no-swap) only on disagreement. The two production callers already
express the same "prefer C-order" intuition; we extend it to the
disagreement case, where the current add-order tie-break has no
layout-friendliness rationale. Cases `enforce_linear_iteration_=true`
would mishandle are exempted by construction (see *Properties preserved*).

## Proposed change

Inside `should_swap(dim0, dim1)`, collect the per-operand verdict (one of
`-1, 0, +1`, or "skipped") for every decidable operand, then decide:

- If all decidable operands agree (non-zero verdicts all equal): keep that
  verdict. This covers same-layout cases including channels-last propagation.
- If decidable operands disagree: return `0` (no-swap), biasing `perm_` toward
  identity (row-major output).
- If only one operand is decidable: keep that operand's verdict (unchanged
  from today).
- If no operand is decidable: return `0` (unchanged from today).

Sketch (aten/src/ATen/TensorIterator.cpp, `reorder_dimensions`):

```cpp
auto should_swap = [&](size_t dim0, size_t dim1) {
  int verdict = 0;            // 0 = no decidable operand yet
  bool have_verdict = false;
  bool disagreement = false;
  for (const auto arg : c10::irange(ntensors())) {
    if (operands_[arg].stride_bytes.empty() || operands_[arg].will_resize)
      continue;
    int64_t stride0 = operands_[arg].stride_bytes[dim0];
    int64_t stride1 = operands_[arg].stride_bytes[dim1];
    if (is_reduction_ && operands_[arg].is_output) {
      // unchanged: reduced-dim push-front takes precedence
      if ((stride0 == 0) != (stride1 == 0)) {
        return stride1 == 0 ? 1 : -1;
      }
    }
    if (stride0 == 0 || stride1 == 0) continue;     // operand undecidable here
    int v;
    if (stride0 < stride1)      v = -1;
    else if (stride0 > stride1) v =  1;
    else {                                            // stride tie: shape tie-break
      // Verbatim from current impl: only t_dim0 > t_dim1 produces a verdict
      // (return 1). On t_dim0 <= t_dim1 the original `continue`s to the next
      // operand -- shape is *not* a symmetric tie-breaker, just a one-sided
      // preference. We keep that semantics: only the > branch emits a
      // verdict; otherwise this operand yields no verdict and we move on.
      auto t_dim0 = shape_[dim0];
      auto t_dim1 = shape_[dim1];
      if (t_dim0 > t_dim1) v = 1;
      else continue;                                  // no verdict, next operand
    }
    if (!have_verdict) {
      verdict = v;
      have_verdict = true;
    } else if (v != verdict) {
      disagreement = true;
      break;
    }
  }
  if (disagreement) return 0;                         // ← the new rule
  return have_verdict ? verdict : 0;
};
```

The reduction branch and the stride/dim-size tie-breaks are kept verbatim from
the current implementation; only the outer aggregation changes.

## Properties preserved by construction

1. **Unary ops**: only one operand. No disagreement possible. Behavior
   identical to today — `reorder_dimensions` still reorders to follow the
   operand's stride, giving dense traversal for permuted views.

2. **In-place ops** (`a.add_(b)`, `a.relu_()` where output aliases an input):
   the aliased operand contributes the same verdict as the input it aliases,
   so disagreement can only come from the non-aliased operand. `should_swap`
   still picks the dominant input's layout when there is no genuine
   disagreement. The output is not re-allocated (its strides are already
   fixed), so the `perm_` only determines traversal order, not output
   layout — the rule changes nothing material here.

3. **Same-layout multi-operand ops** (including channels-last propagation):
   all decidable operands have the same stride ordering, so they agree on
   every dim pair. No disagreement → no row-major bias → layout propagates
   as today. NHWC + NHWC stays NHWC; NCHW-contiguous + NCHW-contiguous stays
   NCHW.

4. **Reduction path**: the `is_reduction_` branch runs first and returns
   directly, bypassing the new aggregation logic.

5. **`fast_set_up` / `NON_OVERLAPPING_DENSE` fast path**: untouched. This
   proposal only modifies the slow path inside `reorder_dimensions`.

6. **`out=` / user-allocated outputs**: untouched. When an output is already
   defined (`!will_resize`), `allocate_or_resize_outputs` does not reallocate;
   the output strides are respected as today.

7. **NumPy output-stride semantics** (Note [Computing output strides]): the
   precedence rule is relaxed *only* on genuine disagreement between decidable
   operands. In agree / single-decidable / undecidable cases the existing rule
   applies verbatim.

## Cases the proposal does and does not solve

**Solved** (output becomes row-major):
- two operands with different non-identity perms (disagreement on dim pairs)
  → row-major output, restoring `a + b == b + a` at the layout level

**Not solved, by design**:
- Both operands agree on a non-row-major perm (e.g. NHWC + NHWC) → output
  stays NHWC. This is layout propagation and is preserved on purpose.
- Single non-row-major operand (unary / in-place) → output follows the
  operand's layout. The rule does not force row-major here, because doing so
  would be a strict regression (it would be `enforce_linear_iteration_=true`,
  which the community has already chosen to default off).

The proposal's value is bounded to "operands disagree on layout" cases. It
strictly never makes things worse than today and strictly improves the
disagreement cases above.

## Risks / open questions

1. **Channels-last in mixed-layout cases.** When a channels-last operand
   meets a row-major operand of the same shape, the two disagree, and the
   rule pulls the output back to row-major. This is a deliberate trade-off
   toward row-major output. We need to sweep `test_view_ops.py`,
   `test_binary_ufuncs.py`, `test_ops.py`, and the channels-last test sweep
   to bound the regression surface before landing. The rule is *softer* than
   `enforce_linear_iteration_=true` (it does not affect same-layout chains)
   but it is not zero-impact on mixed-layout cases.

2. **Implicit `output.strides() == first_input.strides()` assumptions.** Some
   user code and some tests may rely on the current add-order tie-break. The
   disagreement-gated rule changes behavior only when a later operand
   disagrees with the first — which is exactly the case the proposal targets,
   but it is a behavior change and must be validated against the test suite.

3. **Per-dim-pair vs. global verdict.** The sketch above is per-dim-pair: it
   decides each swap independently. An alternative would be to compute the
   full `perm` under each candidate dominant operand and pick the one closest
   to identity. The per-dim-pair version is cheaper and sufficient for the
   cases above; the global version is more aggressive. This proposal chooses
   per-dim-pair to minimize the change surface.

4. **Production callers of `enforce_linear_iteration_` already exist**
   (`nonzero_out_cpu`, `masked_scatter_`), but both set the flag explicitly
   because their downstream stubs require C-order traversal. The
   disagreement-gated proposal does not affect them (they bypass
   `should_swap` entirely via the early return at L247). The proposal
   extends the same "prefer C-order" intuition to general elementwise ops
   *only* in the disagreement case, where the current add-order tie-break
   has no layout-friendliness rationale.

## Microbenchmark plan

```python
import torch, timeit
def bench(a, b, n=100):
    for _ in range(10): _ = a + b
    return timeit.timeit(lambda: a + b, number=n) / n
def chain(a, b): return torch.relu(a + b)

# Before/after comparison:
# bench(tensor_0, tensor_1)        current: non-dense operand wins by add order
# bench(tensor_1, tensor_0)        current: dense operand first, already fine
# chain(tensor_0, tensor_1)        downstream relu gets non-row-major input today
```

Expectation: `chain(tensor_0, tensor_1)` improves after the change (downstream
relu receives a row-major input), the `bench` micro cases may shift modestly,
the same-layout cases (`bench(nhwc_a, nhwc_b)`, `relu_(permuted_view)`) are
unchanged. Numbers will be attached if there is interest in taking this
forward.

## Ask

Before opening a PR, I'd like feedback from layout / TensorIterator
maintainers on:

- Is "disagreement-gated row-major preference" an acceptable relaxation of
  `enforce_linear_iteration_` in principle?
- Is the per-dim-pair formulation the right granularity, or should the
  global-verdict alternative be considered?
- Are there known workloads or tests beyond the ones listed under *Risks*
  that would help bound the regression surface before drafting the
  implementation?

cc @jamesr66a
